README
NAME
    Net::SMTP::Bulk - NonBlocking batch SMTP using Net::SMTP interface

VERSION
    Version 0.19

SYNOPSIS
    This is a rewrite of Net::SMTP using AnyEvent and Coro as a backbone. It
    supports AUTH, SSL and STARTTLS as well. This module can be used as a
    drop in replacement for Net::SMTP. At this point this module is
    EXPIREMENTAL, so use at your own risk. Functionality can change at any
    time.

        use Net::SMTP::Bulk;

        my $smtp = Net::SMTP::Bulk->new($server, %options);

    See Net::SMTP for syntax.

SUBROUTINES/METHODS
  new($server,%options)
  new(%options)
    Options: Host - Hostname or IP address

    Mode - Options of AnyEvent or Coro (default: Coro but switches to
    AnyEvent if Coro is not installed)

    Port - The port to which to connect to on the server (default: 25)

    Hello - The domain name you wish to connect to (default: [same as
    server])

    Debug - Debug information (0-10 depending on level) (default: 0
    [disabled]) OPTIONAL

    DebugPath - Set to default Debug Path. use [HOST] and [THREAD] for
    deeper control of output OPTIONAL

    Secure - If you wish to use a secure connection. (0 - None, 1 - SSL [no
    verify], 1 - SSL [verify], 3 - STARTTLS [no verify], 4 - STARTTLS
    [verify]) OPTIONAL [Requires Net::SSLeay]

    Threads - How many concurrent connections per host (default: 2) OPTIONAL

    Encode - Encode socket( 1: utf8 )

    Callbacks - You can supply callback functions on certain conditions,
    these conditions include:

    connect_pass,connect_fail,auth_pass,auth_fail,reconnect_pass,reconnect_f
    ail,pass,fail,hang

    The callback must return 1 it to follow proper proceedures. You can
    overwrite the defaults by supplying a different return.

    1 - Default

    101 - Remove Thread permanently

    102 - Remove thread temporarily and reconnect at end of batch

    103 - Remove thread temporarily and restart at end of batch (If your
    using an SMTP server with short timeout, it is suggested to use this
    over reconnect)

    104 - Remove Thread temporarily

    202 - Reconnect now

    203 - Restart now

  new(%options, Hosts=>[\%options2,\%options3])
    You can supply multiple hosts in an array.

  auth( [ MECHANISM,] USERNAME, PASSWORD  )
    *Requires Authen::SASL

  mail( ADDRESS )
  to( ADDRESS )
  data()
  datasend( DATA )
  dataend( DATA )
  reconnect(  )
  quit(  )
AUTHOR
    KnowZero

BUGS
    Please report any bugs or feature requests to "bug-net-smtp-bulk at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-SMTP-Bulk>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Net::SMTP::Bulk

    You can also look for information at:

    *   RT: CPAN's request tracker (report bugs here)

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-SMTP-Bulk>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Net-SMTP-Bulk>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Net-SMTP-Bulk>

    *   Search CPAN

        <http://search.cpan.org/dist/Net-SMTP-Bulk/>

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT
    Copyright 2013 KnowZero.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See <http://dev.perl.org/licenses/> for more information.



Changes
Revision history for Perl module Net::SMTP::Bulk

0.19 2014-12-16
    - Added STARTTLS support.
    - Close Debug FH when done for Net::SMTP::Bulk::AnyEvent.
    - Net::SMTP::Bulk::Coro is now partly converted to new debug format (Debug 1 - 10). **THIS BREAKS COMPATIBILITY WITH PREVIOUS VERSIONS**

0.18 2014-11-07
    - Added debug email counter and timestamp to Net::SMTP::Bulk::AnyEvent.
    - Updated some hangs to be quicker to Net::SMTP::Bulk::AnyEvent.
    - Fixed freeze if no email is sent on quit command on Net::SMTP::Bulk::AnyEvent.
    - Fixed proper disconnect on quit command on Net::SMTP::Bulk::AnyEvent.
    - Fixed reconnect to actually work properly on Net::SMTP::Bulk::AnyEvent.
    - Added retry attempts for hang (default: 1), fail (default: 0), global hang (default: 1) to Net::SMTP::Bulk::AnyEvent.

0.17 2014-11-03

    - Fixed issue of handles not being closed correctly in Net::SMTP::Bulk::AnyEvent and fixed some session errors.
    - Fixed improper header issue in Net::SMTP::Bulk::AnyEvent.
    - Added Encode to require list.
    - Added timeout (set to 60 seconds by default), timeout handles and global timeout (set to 120 seconds by default) and global timeout handle.
    - Removed Debug=1 and Debug=2 for Net::SMTP::Bulk::AnyEvent
    - Added Debug 1 - 10 for controlling details of debugging. DebugPath will force writing to filehandle, without setting DebugPath it will print to STDOUT. Setting DebugPath to 'default' will use default path and format. (This breaks compatibility between Net::SMTP::Bulk::AnyEvent and Net::SMTP::Bulk::Coro debugging for now)


0.16 2014-01-29

    - Added new mode for using AnyEvent only.
    - Moved library to Net::SMTP::Bulk::Coro.
    - Added library Net::SMTP::Bulk::AnyEvent

0.15 2013-10-17

    - Fixed warnings when Encode is not used

0.14 2013-10-08

    - Added Encode socket to utf8 option
    - Added Encode write log file to utf8 and set to default

0.13 2013-09-20

    - Fixed order repeat on restart 

0.12 2013-09-19

    - Prevent multiple restarts on multiple fails 

0.11 2013-09-19

    - Changed 102 to be delayed reconnect callback response, 202 for reconnect now
    - Added delayed restart callback response 103, 203 for restart now (restart will close all connections and restart)
    - Fixed remove thread to function correctly and added 104, remove thread temporarily

0.10 2013-09-19

    - Added parallel ssl connect 

0.09 2013-09-18

    - Added parallel connect 

0.08 2013-09-18

    - Fixed auto remove thread issue
    - Added Debug=2 and DebugPath for writing debug data to files
    - added missing reconnect_pass callback response

0.07 2013-09-12

    - Made it work faster(especially on longer response servers) and more efficient by removing coro sleep in favor of readable on the handle

0.06 2013-09-10

    - Fixed to work with Perl 5.10 and below

0.05 2013-09-08

    - Fixed default callbacks to being 1

0.04 2013-09-08

    - Fixed removal of incorrect thread
    - Changed callback status codes to be 3 digits,
      made 1 to be default callback from 0
    - Adjusted reconnect to skip current

0.03 2013-09-08

    - Fixed 220 response causing removal of thread
    - Fixed a few documentation issues

0.02 2013-09-08

    - Fixed <CR><LF>.<CR><LF>
    - Fixed Remove Thread
    - Fixed pass callbacks

0.01 2013-09-07

    - First version, released on an unsuspecting world.



Build.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Net::SMTP::Bulk',
    license             => 'Perl_5',
    dist_author         => q{KnowZero},
    dist_version_from   => 'lib/Net/SMTP/Bulk.pm',
    release_status      => 'stable',
    configure_requires => {
        'Module::Build' => 0,
    },
    build_requires => {
        'Test::More' => 0,
    },
    requires => {
        'AnyEvent'              => 0,
        'Encode'              => 0,
    },
    recommends => {
        'Coro'              => 0,
        'IO::Socket::SSL'              => 0,
        'Authen::SASL'              => 0,
        'MIME::Base64'              => 0,
    },
    add_to_cleanup     => [ 'Net-SMTP-Bulk-*' ],
    create_readme => 1,
    create_makefile_pl => 'traditional',
);

$builder->create_build_script();


lib/Net/SMTP/Bulk.pm
package Net::SMTP::Bulk;

use 5.006;
use strict;
use warnings FATAL => 'all';

#use Encode;
#use Coro;
#use Coro::Handle;
#use AnyEvent::Socket;


=head1 NAME

Net::SMTP::Bulk - NonBlocking batch SMTP using Net::SMTP interface

=head1 VERSION

Version 0.19

=cut

our $VERSION = '0.19';


=head1 SYNOPSIS

This is a rewrite of Net::SMTP using AnyEvent and Coro as a backbone. It supports AUTH, SSL and STARTTLS as well. This module can be used as a drop in replacement for Net::SMTP. At this point this module is EXPIREMENTAL, so use at your own risk. Functionality can change at any time.

    use Net::SMTP::Bulk;

    my $smtp = Net::SMTP::Bulk->new($server, %options);

    
See Net::SMTP for syntax.
    
=head1 SUBROUTINES/METHODS

=head2 new($server,%options)

=head2 new(%options)

Options:
Host - Hostname or IP address

Mode - Options of AnyEvent or Coro (default: Coro but switches to AnyEvent if Coro is not installed)

Port - The port to which to connect to on the server (default: 25)

Hello - The domain name you wish to connect to (default: [same as server])

Debug - Debug information (0-10 depending on level) (default: 0 [disabled]) OPTIONAL

DebugPath - Set to default Debug Path. use [HOST] and [THREAD] for deeper control of output OPTIONAL

Secure - If you wish to use a secure connection. (0 - None, 1 - SSL [no verify], 1 - SSL [verify], 3 - STARTTLS [no verify], 4 - STARTTLS [verify]) OPTIONAL [Requires Net::SSLeay]

Threads - How many concurrent connections per host (default: 2) OPTIONAL

Encode - Encode socket( 1: utf8 )

Callbacks - You can supply callback functions on certain conditions, these conditions include:

connect_pass,connect_fail,auth_pass,auth_fail,reconnect_pass,reconnect_fail,pass,fail,hang

The callback must return 1 it to follow proper proceedures. You can overwrite the defaults by supplying a different return.

1 - Default

101 - Remove Thread permanently

102 - Remove thread temporarily and reconnect at end of batch

103 - Remove thread temporarily and restart at end of batch (If your using an SMTP server with short timeout, it is suggested to use this over reconnect)

104 - Remove Thread temporarily

202 - Reconnect now

203 - Restart now

=head2 new(%options, Hosts=>[\%options2,\%options3])

You can supply multiple hosts in an array.


=head2 auth( [ MECHANISM,] USERNAME, PASSWORD  )

*Requires Authen::SASL

=head2 mail( ADDRESS )

=head2 to( ADDRESS )

=head2 data()

=head2 datasend( DATA )

=head2 dataend( DATA )

=head2 reconnect(  )

=head2 quit(  )

=cut

sub new {
    my $class=shift;
    my %new=@_;
    my $self={};

    if (($new{Mode}||'') eq 'AnyEvent') {
        require Net::SMTP::Bulk::AnyEvent;    
        $self=Net::SMTP::Bulk::AnyEvent->new(@_);    
    } else {
        if (eval { require Net::SMTP::Bulk::Coro; 1 }) {
            $self=Net::SMTP::Bulk::Coro->new(@_);
        } else {
            require Net::SMTP::Bulk::AnyEvent; 
            $self=Net::SMTP::Bulk::AnyEvent->new(@_);
        }
    
        
    }
    
    

    return $self;
}


#########################################################################


=head1 AUTHOR

KnowZero

=head1 BUGS

Please report any bugs or feature requests to C<bug-net-smtp-bulk at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-SMTP-Bulk>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::SMTP::Bulk


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-SMTP-Bulk>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Net-SMTP-Bulk>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Net-SMTP-Bulk>

=item * Search CPAN

L<http://search.cpan.org/dist/Net-SMTP-Bulk/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 KnowZero.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Net::SMTP::Bulk
